README.md
# abcmusiex
ABCMusiEx generates sight reading exercises: ABCs, printable PS/PDFs, audio WAV/MP3. Student read and practice generated scores while abcmusiex backing track follows at different speeds. Highly configurable and easy.


doc/ChangeLog
* Mon May 1 2013 Rodolfo Alcazar <rodolfoap@gmail.com> 1.0-9
- Merging ABCMidi needed modules due to the lack of Fedora RPM

* Wed Jul 1 2010 Rodolfo Alcazar <rodolfoap@gmail.com> 1.0-8
- Fixes: minor syscommand fixes

* Wed Jun 23 2010 Rodolfo Alcazar <rodolfoap@gmail.com> 1.0-7
- Fifth release signed 1.0-7
- Fixes: minor bugs in system command
-        Spec file with some Willy Cat <siamwilly@yahoo.com.ar> suggestions
- Adds: Graphic user interface

* Tue Jun 17 09:02:17 2008 Rodolfo Alcazar <rodolfoap@gmail.com>
- Fifth release signed 0.9-5
- Fixes: using yaps instead of abcm2ps

* Fri Dec 28 09:44:37 2007 Rodolfo Alcazar <rodolfoap@gmail.com>
— Fourth release signed 0.9-2
- Fixes: A couple of bugs
-        Uses evince instead of acroread, timidity as player, not needed wav
	 or mp3 files for exercises starting
-        Faster generation and action
-        Fedora 8 ready, avoid many dependences, rpm process generation improved
-        Stages generation improved
-        Pulseaudio not fixed in timidity, but abcmusiex now working fine with
	 default device.

* Tue Apr  4 16:00:00 2006 Rodolfo Alcazar <rodolfoap@gmail.com>
— Third release signed 0.8-2
- Fixes: Generation algorithms improved;
-        Sharps and flats generator;
-        Ties generator.

* Thu Mar 23 15:00:00 2006 Rodolfo Alcazar <rodolfoap@gmail.com>
— Second release signed 0.8
- Fixes: Controlling generation stages;
-        Updated documentation and manpage;
-        Better generation tools;
-        Metronom optional;
-        Randomizer bug fixed;
-        Dispersion fixed;
-        Interface standardization.

* Mon Aug 22 16:00:00 2005 Rodolfo Alcazar <rodolfoap@gmail.com>
— First release signed 0.7.



src/abcmusiex.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "definitions.h"
#include "abcmusiex.h"
#include "functions.h"
#include "time.h"

#define  NOTNORMPROB 12
#define  NOTDOTPROB 8
#define  NOTTRPPROB 8
#define  NOTRSTPROB 6
#define  SFUNIVERSE 6
/* Sharps and flats universe 0=b,1=#,2,3=2,4=2,5=2*/
#define  LEUNIVERSE 6
/* Legato universe one of eight is legato */

char *notes[]={"C,","D,","E,","F,","G,","A,","B,","C","D","E","F","G","A","B","c","d","e","f","g","a","b","c'","d'","e'","f'","g'","a'","b'", "c''"};
char *notesS[]={"^C,","^D,","E,","^F,","^G,","^A,","B,","^C","^D","E","^F","^G","^A","B","^c","^d","^e","f","^g","^a","b","^c'","^d'","e'","^f'","^g'","^a'","b'", "^c''"};
char *notesF[]={"C,","_D,","_E,","F,","_G,","_A,","_B,","C","_D","_E","F","_G","_A","_B","c","_d","_e","f","_g","_a","_b","c'","_d'","_e'","f'","_g'","_a'","_b'", "c''"};
//    notes array goes from 0 to 28

/*--------------------------------------------------------------------------------------------------------------------------------------------------------*/
/** Process starts with this function: generates a bar based on metrics:  
  * 4/4 --> [4][4][4][4].
  *                                                                                                                                                       */
void setbarbeats(int beatsperbar, int beatlongbar, char* barbuffer){
	int count;
	char unit[]="[--]";
	strcpy(barbuffer, "");
	if(beatlongbar==8) strcpy(unit, "[8]");
	if(beatlongbar==4) strcpy(unit, "[4]");
	if(beatlongbar==2) strcpy(unit, "[2]");
	for(count=0;count<beatsperbar;count++)
		strcat(barbuffer, unit);
}

/*--------------------------------------------------------------------------------------------------------------------------------------------------------*/
/** Then the metrics-generated bar is compressed: [4][4][4][4] --> [1] to be splitted on several and pseudo-random notes.
  * 
  *                                                                                                                                                       */
void compressbar(char* barbuffer){
	while(strlen(barbuffer)>6){
		replace(barbuffer, "[8][8]", "[4]");
		replace(barbuffer, "[4][4]", "[2]");
		replace(barbuffer, "[2][2]", "[1]");
	}
	replace(barbuffer, "[8][8]", "[4]");
	replace(barbuffer, "[4][4]", "[2]");
	replace(barbuffer, "[2][2]", "[1]");
	replace(barbuffer, "[2][4]", "[4] [4] [4]");
}

/*--------------------------------------------------------------------------------------------------------------------------------------------------------*/
/** Then, the starting bar is marked as divisible: [1] --> [1d]
  *  
  *                                                                                                                                                       */
void genrtbar(char* barbuffer, char* barmodel){
	strcpy(barbuffer, barmodel);
	replace(barbuffer, "]", "d]");
}

/*--------------------------------------------------------------------------------------------------------------------------------------------------------*/
/** Now the main routine: splitting barbuffer randomicly until split max level is reached: 
  * INIT --> [1d] 
  * [1d] --> [2d] [2d]
  * [1d] --> [4d] [4d]     [4d] [4d]
  * [1d] --> [4]  [8d][8d] [4]  [8d][8d]
  * [1d] --> [4]  [8][8]   [4]  [16][16][8]
  *                                                                                                                                                       */
void splitbar(char* barbuffer, int splitlevel, int splitmlevel){
	if(splitlevel<=splitmlevel){
		if(splitlevel==1) {
			replace_prob(barbuffer, "[1d]", "[1]", NOTNORMPROB*splitlevel);
			if(RESTNTES==1)replace_prob(barbuffer, "[1d]", " [1z] ", NOTRSTPROB*splitlevel);
			if(RESTNTES==1)replace_prob(barbuffer, "[1d]", " [2z][2d] ", NOTRSTPROB*splitlevel);
			if(RESTNTES==1)replace_prob(barbuffer, "[1d]", " [2d][2z] ", NOTRSTPROB*splitlevel);
			if(BROKENTS==1)replace_prob(barbuffer, "[1d]", " [2]>[2] ", NOTDOTPROB*(splitlevel-1));
			if(BROKENTS==1)replace_prob(barbuffer, "[1d]", " [2]<[2] ", NOTDOTPROB*(splitlevel-1));
			if(TRIPLETS==1)replace_prob(barbuffer, "[1d]", " (3[2][2][2] ", NOTTRPPROB*(splitlevel-1));
			if(splitlevel==splitmlevel) {
				replace (barbuffer, "[1d]", "[2] [2]");
				replace (barbuffer, "[2d]", "[2]");
			}
			else
				replace (barbuffer, "[1d]", "[2d] [2d]");
		}
		if(splitlevel==2) {
			replace_prob(barbuffer, "[2d]", "[2]", NOTNORMPROB*splitlevel);
			if(RESTNTES==1)replace_prob(barbuffer, "[2d]", " [2z] ", NOTRSTPROB*splitlevel);
			if(RESTNTES==1)replace_prob(barbuffer, "[2d]", " [4z][4d] ", NOTRSTPROB*splitlevel);
			if(RESTNTES==1)replace_prob(barbuffer, "[2d]", " [4d][4z] ", NOTRSTPROB*splitlevel);
			if(BROKENTS==1)replace_prob(barbuffer, "[2d]", " [4]>[4]", NOTDOTPROB*(splitlevel-1));
			if(BROKENTS==1)replace_prob(barbuffer, "[2d]", " [4]<[4]", NOTDOTPROB*(splitlevel-1));
			if(TRIPLETS==1)replace_prob(barbuffer, "[2d]", " (3[4][4][4] ", NOTTRPPROB*(splitlevel-1));
			if(splitlevel==splitmlevel) {
				replace (barbuffer, "[2d]", "[4] [4]");
				replace (barbuffer, "[4d]", "[4]");
			}
			else
				replace (barbuffer, "[2d]", "[4d] [4d]");
		}
		if(splitlevel==3) {
			replace_prob(barbuffer, "[4d]", "[4]", NOTNORMPROB*splitlevel);
			if(RESTNTES==1)replace_prob(barbuffer, "[4d]", " [4z] ", NOTRSTPROB*splitlevel);
			if(RESTNTES==1)replace_prob(barbuffer, "[4d]", " [8z][8d] ", NOTRSTPROB*splitlevel);
			if(RESTNTES==1)replace_prob(barbuffer, "[4d]", " [8d][8z] ", NOTRSTPROB*splitlevel);
			if(BROKENTS==1)replace_prob(barbuffer, "[4d]", " [8]>[8] ", NOTDOTPROB*(splitlevel-1));
			if(BROKENTS==1)replace_prob(barbuffer, "[4d]", " [8]<[8] ", NOTDOTPROB*(splitlevel-1));
			if(TRIPLETS==1)replace_prob(barbuffer, "[4d]", " (3[8][8][8] ", NOTTRPPROB*(splitlevel-1));
			if(splitlevel==splitmlevel) {
				replace (barbuffer, "[4d]", "[8][8]");
				replace (barbuffer, "[8d]", "[8]");
			}
			else
				replace (barbuffer, "[4d]", "[8d][8d]");
		}
		if(splitlevel==4) {
			replace_prob(barbuffer, "[8d]", "[8]", NOTNORMPROB*splitlevel);
			if(RESTNTES==1)replace_prob(barbuffer, "[8d]", " [8z] ", NOTRSTPROB*splitlevel);
			if(RESTNTES==1)replace_prob(barbuffer, "[8d]", " [16z][16d] ", NOTRSTPROB*splitlevel);
			if(RESTNTES==1)replace_prob(barbuffer, "[8d]", " [16d][16z] ", NOTRSTPROB*splitlevel);
			replace     (barbuffer, "[8d]", "[16][16]");
			replace     (barbuffer, "[16d]", "[16]");
		}
		replace (barbuffer, "  ", " ");
		replace (barbuffer, "  ", " ");
		replace (barbuffer, "  ", " ");
		replace (barbuffer, "  ", " ");
		splitbar(barbuffer, splitlevel+1, splitmlevel);
	}
}

/*--------------------------------------------------------------------------------------------------------------------------------------------------------*/
/** Now, some optimizations. This code is not completed
  *                                                                                                                                                       */
void checkbar(char* barbuffer){
	//replace(barbuffer, "[8][8][8] [8][8] [8][8]", " [8][8][8] [8][8][8][8]");
	//replace(barbuffer, "[8][8] [8][8] ", "[8][8][8][8] ");
	//replace(barbuffer, " [8][8] [8][8]", " [8][8][8][8]");
}

/*--------------------------------------------------------------------------------------------------------------------------------------------------------*/
/** Generating notes based on a dispersion from an original note
  *                                                                                                                                                       */
int disperse_note(int note, int dispersion, int minnote, int maxnote){
	if(ONLEGATO==0)	{
		note+=random_num(dispersion*2+1) - dispersion;
        if(note<minnote)note=minnote+dispersion;
        if(note>maxnote)note=maxnote-dispersion;
	}
	
    return note;
}

/*--------------------------------------------------------------------------------------------------------------------------------------------------------*/
/** The vertical approach just replaces generated bar with notes within a range and a dispersion factor:
  * [4] [8][8] [4] [16][16][8] --> c d/2b/2 d a/4f/4b/2
  * The generated bar is ready to be appended at the end of the sngbuffer
  *                                                                                                                                                       */
int  putnotes(char* barbuffer, int note){
	int status;

	replace(barbuffer, "[16]", "[n]/4");
	replace(barbuffer, "[8]",  "[n]/2");
	replace(barbuffer, "[4]",  "[n]");
	replace(barbuffer, "[2]",  "[n]2");
	replace(barbuffer, "[1]",  "[n]4");

	replace(barbuffer, "[16z]", "z/4");
	replace(barbuffer, "[8z]",  "z/2");
	replace(barbuffer, "[4z]",  "z");
	replace(barbuffer, "[2z]",  "z2");
	replace(barbuffer, "[1z]",  "z4");

	while((status=replacenotes(barbuffer, note))) 
		note=disperse_note(note, DISPERSION, MINNOTE, MAXNOTE);
	
	return note;
}

/*--------------------------------------------------------------------------------------------------------------------------------------------------------*/
/** Here, each note is replaced: [n] --> ^c) (c sharp closing legato)
  *  
  *                                                                                                                                                       */
int replacenotes(char* barbuffer, int note){
	int notype, status;
	char notstr[9]; /* notstr = notesF[note] = "_C," or whatever structure is*/
	int  ONLEGATO_TEMP;
	
	ONLEGATO_TEMP=ONLEGATO;

/* Sharps and flats generation */
	status=0;
	notype=random_num(SFUNIVERSE); 
	/* notype=0-b,1-#,2,3,4,5 */
 	if(notype>2) notype=2;
	/* notype=0-b,1-#,2,2,2,2 */
	if(GENFLATS==0 && notype==0) notype=2;
	/* notype=  2,1-#,2,2,2,2 */
	if(GENSHRPS==0 && notype==1) notype=2;
	/* notype=0-b,2  ,2,2,2,2 */
	
	if(ONLEGATO==1) notype=2;
	
	/* Put sharps or/and flats*/
	if(notype==2) strcpy(notstr, notes[note]);
	if(notype==1) strcpy(notstr, notesS[note]);
	if(notype==0) strcpy(notstr, notesF[note]);

/* Ties/legato generation */
	if(GENLEGAT==1) {
		if(ONLEGATO==1){
			ONLEGATO=0;
		}
		else {
			if(random_num(LEUNIVERSE)==0) {
				strcat(notstr, "-");
				ONLEGATO=1;
			}	
		}
	}
	
	status=replace_once(barbuffer, "[n]", notstr);	
	//replace(barbuffer, "- ", "-");	
	replace(barbuffer, "-/", "/-");	
	replace(barbuffer, "-2", "2-");	
	replace(barbuffer, "-4", "4-");	
	replace(barbuffer, "-8", "8-");	
	
	if(!status)	ONLEGATO=ONLEGATO_TEMP;

	return status;
}

/*--------------------------------------------------------------------------------------------------------------------------------------------------------*/
/** Then, the starting bar is marked as divisible: [1] --> [1d]
  *  
  *                                                                                                                                                       */
void genrtclk(char* barbuffer, char* barmodel){
	strcpy(barbuffer, barmodel);
	replace(barbuffer, "[4]", "A");
	replace(barbuffer, "[2]", "A A");
	replace(barbuffer, "[1]", "A A A A");
}

/*--------------------------------------------------------------------------------------------------------------------------------------------------------*/
/** Main process. It is not called main cause it might be included in a glade project.
  * 
  *                                                                                                                                                       */
void generate(void) {
	/* Main definitions */
	char *barmodel, *barbuffer, *rstbuffer, *hdrbuffer, *sngbuffer,  *filbuffer;
	barmodel =(char *)malloc(BARSIZEMAX);
	barbuffer=(char *)malloc(BARSIZEMAX);
	rstbuffer=(char *)malloc(BARSIZEMAX);
	hdrbuffer=(char *)malloc(HDRSIZEMAX);
	sngbuffer=(char *)malloc(SNGSIZEMAX);
	filbuffer=(char *)malloc(FILSIZEMAX);

	int count=0, note=INITIALNOTE;

	srand((unsigned)time(NULL));

	/* Get the beats quantity, so we can generate a bar with this size */
	setbarbeats(BEATSPERBAR, BEATLONGBAR, barbuffer);
	compressbar(barbuffer);

	strcpy(barmodel, barbuffer);

	/* Generate header and initialize song buffer with it */
	genrthdr(hdrbuffer);
	strcpy(sngbuffer, hdrbuffer);

	genrtrst(rstbuffer);
	strcat(sngbuffer, rstbuffer);
	strcat(sngbuffer, BARSEPRTR);

	/* Then we start the count */
	for(count=0; count<BARSPERSONG-1; count++){
		genrtbar(barbuffer, barmodel);
		splitbar(barbuffer, 1, SPLITMLEVEL);
		checkbar(barbuffer);
		
		note=putnotes(barbuffer, note);

		strcat(sngbuffer, barbuffer);
		strcat(sngbuffer, BARSEPRTR);
	}
	strcat(sngbuffer, BARENDING);
	strcat(sngbuffer, "\n");

	/* Generate metro */
	if(GENMETRO==1){
		genrtmtr(hdrbuffer);
		strcat(sngbuffer, hdrbuffer);
		for(count=0; count<BARSPERSONG; count++){
			genrtclk(barbuffer, barmodel);
			strcat(sngbuffer, barbuffer);
			strcat(sngbuffer, BARSEPRTR);
		}
		strcat(sngbuffer, BARENDING);
	}

	strcpy(filbuffer, sngbuffer);
	wrtfil(filbuffer);
	free(barmodel);
	free(barbuffer);
	free(rstbuffer);
	free(hdrbuffer);
	free(sngbuffer);
	free(filbuffer);
}

/*--------------------------------------------------------------------------------------------------------------------------------------------------------*/
/** Main process, oriented mainly to arguments processing
  * 
  *                                                                                                                                                       */
int  main (int argc, char *argv[ ]){
	int argn;

	strcpy(ARGTEMPO, "60");
	strcpy(ARGTITLE, "Music Score Reading Exercise");
	strcpy(ARGBEATS, "4");
	strcpy(ARGNOTEWID, "4");
	strcpy(ARGSONGKEY, "C");
	strcpy(ARGCOMPOSER, "Generated with ABCMusiEx");
	strcpy(ARGWORDS, "This printable and audio parts are generated under the terms of the GNU license. ");
	strcpy(ARGCLEF, "treble");

	BARSPERSONG=162;
	strcpy(ALLFILE, "exercise");

	SPLITMLEVEL=2;
	DISPERSION=4;
	MINNOTE=2;
	MAXNOTE=21;

	TRIPLETS=0;
	BROKENTS=0;
	RESTNTES=0;
	SEEDSTRT=0;

	GENSTAGE=4;
	GENMETRO=1;

	GENSHRPS=0;
	GENFLATS=0;
	GENLEGAT=0;
	ONLEGATO=0;

	for(argn=0; argn<argc; argn++)
	{
		if(strcmp(argv[argn], "--h")==0 || strcmp(argv[argn], "-h")==0 || strcmp(argv[argn], "--help")==0 || argc<=1) {
			printf("GNU 'abcmusiex' generates music score and audio exercises for improving musical score reading.\n\n");
			printf("Usage: abcmusiex [OPTION]...\n\nOptions:\n");
			printf("\t-t TEMPO\ttempo (20-300, default TEMPO=60)\n");
			printf("\t-b BEATS\tbeats (2-4, default BEATS=4), metric numerator\n");
			printf("\t-w WIDTH\twidth (2, 4 or 8, default WIDTH=4), 2vs, 4ths or 8ths, metric denominator\n");
			printf("\t-k KEY\t\tkey, default KEY=C\n");
			printf("\t-c CLEF\t\tclef: treble(default), alto, tenor, baritone, bass, mezzo, soprano, perc\n");
			printf("\t-n QUANTITY\tnumber of bars to generate, default QUANTITY=162\n");
			printf("\t-o FILENAME\toutput filenames, default FILENAME=exercise.*\n");
			printf("\t-l LEVEL\tlevel, only use 1 to 4: default LEVEL=2; 1 is 1/2s, 2 is 4ts, 3 is 8ts, 4 is 16ths\n");
			printf("\t-d DISPERSION\tdispersion (distance from last generated tone, in notes), default DISPERSION=3 notes\n");
			printf("\t-T TITLE\ttitle\n");
			printf("\t-C AUTHOR\tauthor\n");
			printf("\t-W COMMENT\tcomment words\n");
			printf("\t--triplets\tinclude triplets in exercise\n");
			printf("\t--dots\t\tinclude dotted couples\n");
			printf("\t--rests\t\tinclude rests\n");
			printf("\t--sharps\tgenerate random sharps\n");
			printf("\t--flats\t\tgenerate random flats\n");
			printf("\t--legato\tgenerate random legatos\n");
			printf("\t--min I\t\tminnote, default I=2; low-low-C=0; low-C=7; C=14; high-C=21; high-high-C=28\n");
			printf("\t--max A\t\tmaxnote, default A=21; low-low-C=0; low-C=7; C=14; high-C=21; high-high-C=28\n");
			printf("\t--seed S\tseed start, 0...10000, default S=0\n\n");
			printf("\t--stage STAGE\tgeneration stage: 0(abc), 1(+mid), 2(+ps), 3(shows .ps), 4(plays .mid, default), 5(+wav, neither show nor play), 6(+mp3), 7(+pdf)\n");
			printf("\t--nometro\tdoes not generate metronom click\n\n");
			printf("\t-h, --help\tshows this help\n");
			printf("\t--version \tshows version\n\n");
			printf("Examples:\n\n");
			printf("\tabcmusiex --help\t\t\tshows this help\n");
			printf("\tabcmusiex -l 3\t\t\t\tgenerates an 8ths exercise\n");
			printf("\tabcmusiex -t 100\t\t\tgenerates a tempo 100 exercise\n");
			printf("\tabcmusiex -n 10 -c F\t\t\tgenerates 10 measures on F key\n");
			printf("\tabcmusiex -T \"Exercise 1\" -o exercise1\tentitles exercise\n\n");
			printf("Report bugs to <rodolfoap@gmail.com>\n");
			return 0;
		}

		if(strcmp(argv[argn], "--version")==0) {
			printf("ABCMusiEX %s\n\n", VERSOFT);
			printf("Copyright (C) 1999 Free Software Foundation, Inc.\n");
			printf("This is free software; see the source for copying conditions.  There is NO\n");
			printf("warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n\n");
			printf("Written by Rodolfo Alcazar <rodolfoap@gmail.com>\n");
			return 0;
		};
	
		// Is getopt neccesary?? This seems simpler!

		if(strcmp(argv[argn], "-t")==0 && (argn+1) < argc) strcpy(ARGTEMPO, argv[++argn]);
		if(strcmp(argv[argn], "-T")==0 && (argn+1) < argc) strcpy(ARGTITLE, argv[++argn]);
		if(strcmp(argv[argn], "-b")==0 && (argn+1) < argc) strcpy(ARGBEATS, argv[++argn]);
		if(strcmp(argv[argn], "-w")==0 && (argn+1) < argc) strcpy(ARGNOTEWID, argv[++argn]);
		if(strcmp(argv[argn], "-k")==0 && (argn+1) < argc) strcpy(ARGSONGKEY, argv[++argn]);
		if(strcmp(argv[argn], "-C")==0 && (argn+1) < argc) strcpy(ARGCOMPOSER, argv[++argn]);
		if(strcmp(argv[argn], "-W")==0 && (argn+1) < argc) strcpy(ARGWORDS, argv[++argn]);
		if(strcmp(argv[argn], "-c")==0 && (argn+1) < argc) strcpy(ARGCLEF, argv[++argn]);

		if(strcmp(argv[argn], "-n")==0 && (argn+1) < argc) BARSPERSONG=atoi(argv[++argn]);
		if(strcmp(argv[argn], "-o")==0 && (argn+1) < argc) strcpy(ALLFILE, argv[++argn]);

		if(strcmp(argv[argn], "-l")==0 && (argn+1) < argc) SPLITMLEVEL=atoi(argv[++argn]);
		if(strcmp(argv[argn], "-d")==0 && (argn+1) < argc) DISPERSION=atoi(argv[++argn]);
		if(strcmp(argv[argn], "--min")==0 && (argn+1) < argc) MINNOTE=atoi(argv[++argn]);
		if(strcmp(argv[argn], "--max")==0 && (argn+1) < argc) MAXNOTE=atoi(argv[++argn]);
		if(strcmp(argv[argn], "--seed")==0 && (argn+1) < argc) SEEDSTRT=atoi(argv[++argn]);
		if(strcmp(argv[argn], "--stage")==0 && (argn+1) < argc) GENSTAGE=atoi(argv[++argn]);

		if(strcmp(argv[argn], "--triplets")==0) TRIPLETS=1;
		if(strcmp(argv[argn], "--dots")==0) BROKENTS=1;
		if(strcmp(argv[argn], "--rests")==0) RESTNTES=1;
		if(strcmp(argv[argn], "--sharps")==0) GENSHRPS=1;
		if(strcmp(argv[argn], "--flats")==0) GENFLATS=1;
		if(strcmp(argv[argn], "--legato")==0) GENLEGAT=1;
		if(strcmp(argv[argn], "--nometro")==0) GENMETRO=0;
	}
	BEATSPERBAR=atoi(ARGBEATS);
	BEATLONGBAR=atoi(ARGNOTEWID);
	INITIALNOTE=MINNOTE + (MAXNOTE-MINNOTE)/2;

	srand(SEEDSTRT*3+1);

	strcpy(ABCFILE, ALLFILE); strcat(ABCFILE, ".abc");
	strcpy(PSFILE,  ALLFILE); strcat(PSFILE,  ".ps" );
	strcpy(PDFFILE, ALLFILE); strcat(PDFFILE, ".pdf");
	strcpy(MIDFILE, ALLFILE); strcat(MIDFILE, ".mid");
	strcpy(WAVFILE, ALLFILE); strcat(WAVFILE, ".wav");
	strcpy(MP3FILE, ALLFILE); strcat(MP3FILE, ".mp3");

	generate();
	return 0;
	free(ARGTEMPO);
}


src/functions.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "functions.h"
#include "definitions.h"

/*--------------------------------------------------------------------------------------------------------------------------------------------------------*/
/** With a parameter range=100 this function generates numbers in a range 
  * of 0 to 99, meaning 0 < RANDOM_NUMBER < range, being range an integer.
  *                                                                                                                                                       */
int  random_num(int range){
	return (float)rand()/RAND_MAX*range;
}

/*--------------------------------------------------------------------------------------------------------------------------------------------------------*/
/** ABC metronome is generated with this routine
  * 
  *                                                                                                                                                       */
void genrtmtr(char* hdrbuffer){
	int count=0;
	char *metro[]={
	        /* 01 */ "V: 2\n",
	        /* 02 */ "%%MIDI channel 10\n",
	        /* 03 */ "%%MIDI program 10\n",
	        /* 04 */ "%%MIDI transpose -25\n",
	        /* 05 */ "K: C",
	};
	strcpy(hdrbuffer, "");
	for(count=0;count<5;count++) {
		strcat(hdrbuffer, metro[count]);
	}
	strcat(hdrbuffer, "\n");
}

/*--------------------------------------------------------------------------------------------------------------------------------------------------------*/
/** ABC headers are generated with this function into a header buffer: hdrbuffer = "% ABC notation ....."
  * 
  *                                                                                                                                                       */
void genrthdr(char* hdrbuffer){
	int count=0;
	char *header[]={
	        /* 01 */ "% ABC notation for midi generation file",
	        /* 02 */ "\n%%topmargin 80\n% X = Song Index\nX: ",
	        /* 03 */ "1",
	        /* 04 */ "\n% L = Default Lenght\nL: ",
	        /* 05 */ "1/4",
	        /* 06 */ "\n% T = Title\nT: ",
	        /* 07 */ ARGTITLE,
	        /* 08 */ "\n% Q = Tempo\nQ: ",
	        /* 09 */ ARGTEMPO,
	        /* 10 */ "\n% C = Composer\nC: ",
	        /* 11 */ ARGCOMPOSER,
	        /* 12 */ "\n% S = Source\nS: ",
	        /* 13 */ "ABCMusiEx V2",
	        /* 14 */ "\n% Z = Transcriber\nZ: ",
	        /* 15 */ "ABCMusiEx V2",
	        /* 16 */ "\n% M = Meter\nM: ",
	        /* 17 */ ARGBEATS,
	        /* 18 */ "/",
	        /* 18 */ ARGNOTEWID,
		/* 19 */ "\n% K = Key \nK: ",
	        /* 19 */ ARGSONGKEY,
		/* 19 */ " clef=",
	        /* 19 */ ARGCLEF,
	        /* 20 */ "\n% P = Parts, A, A, B, etc.\nP: ",
	        /* 21 */ "A",
	        /* 22 */ "\n% W = Words\nW: ",
	        /* 23 */ ARGWORDS,
	        /* 24 */ "\n% V = Clef\nV: ",
	        /* 25 */ "1"
	};
	strcpy(hdrbuffer, "");
	for(count=0;count<29;count++) {
		strcat(hdrbuffer, header[count]);
	}
	strcat(hdrbuffer, "\n");
}

/*--------------------------------------------------------------------------------------------------------------------------------------------------------*/
/** A routine for replacing all characters, pasted from somewhere at the Internet
  * 
  *                                                                                                                                                       */
int replace(char *string, char *oldpiece, char *newpiece){
   int str_index, newstr_index, oldpiece_index, end, new_len, old_len, cpy_len;
   char *c;
   static char *newstring;

   newstring=(char*)malloc(sizeof(string)*40);

   if ((c = (char *) strstr(string, oldpiece)) == NULL) return 0;

   new_len        = strlen(newpiece);
   old_len        = strlen(oldpiece);
   end            = strlen(string) - old_len;
   oldpiece_index = c - string;

   newstr_index = 0;
   str_index = 0;
   while(str_index <= end && c != NULL)
   {
        cpy_len = oldpiece_index-str_index;

        strncpy(newstring+newstr_index, string+str_index, cpy_len);
        newstr_index += cpy_len;
        str_index    += cpy_len;

        strcpy(newstring+newstr_index, newpiece);
        newstr_index += new_len;
        str_index    += old_len;

        if((c = (char *) strstr(string+str_index, oldpiece)) != NULL) oldpiece_index = c - string;
   }
   strcpy(newstring+newstr_index, string+str_index);
   strcpy(string, newstring);
   free(newstring);
   return 1;
}

/*--------------------------------------------------------------------------------------------------------------------------------------------------------*/
/** A routine for replacing based on probabilities, based on last one
  * 
  *                                                                                                                                                       */
int replace_prob(char *string, char *oldpiece, char *newpiece, int probability) {

   int str_index, newstr_index, oldpiece_index, end, new_len, old_len, cpy_len, r;
   char *c;
   static char *newstring;

   newstring=(char*)malloc(sizeof(string)*40);

   if ((c = (char *) strstr(string, oldpiece)) == NULL) return 0;

   new_len        = strlen(newpiece);
   old_len        = strlen(oldpiece);
   end            = strlen(string) - old_len;
   oldpiece_index = c - string;

   newstr_index = 0;
   str_index = 0;
   while(str_index <= end && c != NULL)
   {
        r=random_num(100);
        cpy_len = oldpiece_index-str_index;

        strncpy(newstring+newstr_index, string+str_index, cpy_len);
        newstr_index += cpy_len;
        str_index    += cpy_len;

        if(r<probability){
                strcpy(newstring+newstr_index, newpiece);
                newstr_index += new_len;
                str_index    += old_len;
        }
        else {
                strcpy(newstring+newstr_index, oldpiece);
                newstr_index += old_len;
                str_index    += old_len;
        }


        if((c = (char *) strstr(string+str_index, oldpiece)) != NULL) oldpiece_index = c - string;
   }
   strcpy(newstring+newstr_index, string+str_index);
   strcpy(string, newstring);
   free(newstring);
   return 1;
}

/*--------------------------------------------------------------------------------------------------------------------------------------------------------*/
/** A routine for replacing once, based on first one
  * 
  *                                                                                                                                                       */
int replace_once(char *string, char *oldpiece, char *newpiece) {

   int str_index, newstr_index, oldpiece_index, end, new_len, old_len, cpy_len;
   char *c;
   static char *newstring;

   newstring=(char*)malloc(sizeof(string)*40);

   if ((c = (char *) strstr(string, oldpiece)) == NULL) return 0;

   new_len        = strlen(newpiece);
   old_len        = strlen(oldpiece);
   end            = strlen(string) - old_len;
   oldpiece_index = c - string;

   newstr_index = 0;
   str_index = 0;
   cpy_len = oldpiece_index-str_index;

   strncpy(newstring+newstr_index, string+str_index, cpy_len);
   newstr_index += cpy_len;
   str_index    += cpy_len;

   strcpy(newstring+newstr_index, newpiece);
   newstr_index += new_len;
   str_index    += old_len;

   if((c = (char *) strstr(string+str_index, oldpiece)) != NULL) oldpiece_index = c - string;
   strcpy(newstring+newstr_index, string+str_index);
   strcpy(string, newstring);
   free(newstring);
   return 1;
}

/*--------------------------------------------------------------------------------------------------------------------------------------------------------*/
/** Write the file buffer to a file
  * 
  *                                                                                                                                                       */
void wrtfil(char* filbuffer){
	char syscommand[100];
	int count=0;
	FILE *fileout;
	char *syscommands[] = {
		"abc2midi %s -o %s -RS", ABCFILE, MIDFILE,
		"abc2abc %s -V 1 -n 6 -e | yaps - -M 50x100 -s 0.8 -o %s", ABCFILE, PSFILE,
		"evince %s %s &", PSFILE, "",
		"(timidity %s; pkill evince) &", MIDFILE, "",
		//"timidity -ig %s &", MIDFILE, "",	// graphical interface
		"timidity %s -Ow -o %s", MIDFILE, WAVFILE,
		"lame %s %s", WAVFILE, MP3FILE,
		"gs -dNOPAUSE -sDEVICE=pdfwrite -dBATCH -sOutputFile=%s %s", PDFFILE, PSFILE
	};
	fileout=fopen(ABCFILE, "w");
	fprintf(fileout, "%s\n", filbuffer);
	fclose(fileout);
	printf("ABC file generated.\n");
	for(count=0; count<GENSTAGE*3; count+=3){
		sprintf(syscommand, syscommands[count], syscommands[count+1], syscommands[count+2]);
		if(GENSTAGE>4 && (count==6 || count==9)) {
			printf("\nSkipping shell command:\n\n$ %s\n\n", syscommand);
		}
		else {
			printf("\nExecuting shell command:\n\n$ %s\n\n", syscommand);
			system(syscommand);
		}
	}
}

/*--------------------------------------------------------------------------------------------------------------------------------------------------------*/
/** Initial song rests are generated with this routine. Very simple: barbuffer = "Z"
  *
  *                                                                                                                                                       */
void genrtrst(char* barbuffer){
        strcpy(barbuffer, "Z");
}
